R/data.table.R

```diff
@@ -1891,10 +1891,15 @@ as.matrix.data.table <- function(x, rownames, ...) {
   rn <- NULL
   rnc <- NULL
   if (!missing(rownames)) { # Convert rownames to a column index if possible
-    if (length(rownames) == nrow(x)) {
+    if (length(rownames) == nrow(x) && nrow(x) > 1) {
       # rownames argument is a vector of row names, no column in x to drop.
       rn <- rownames
       rnc <- NULL
+    } else if (length(rownames) == nrow(x) && nrow(x) == 1 && is.character(rownames) && !(rownames %in% colnames(x))) {
```

Comment:
I think an extra argument for the length-1 case is going to be necessary. Something like `rownames.literal = TRUE`. 

```r
as.matrix(data.table(x = 1, z = 2), "z")
#>   x z
#> z 1 2
as.matrix(data.table(x = 1, z = 2), "z", rownames.literal = FALSE)
#>   x z
#> 2 1 2
```

Reply:
Maybe. I'm not sure having an extra argument is really necessary, since it will only apply to data.tables with 1 row, unless we want to allow vector recycling for the rownames when the data.table has > 1 rows.

Reply:
But this PR arose because of confusion about 1-row data.tables. And there may be other cases where the overloaded argument of `rownames` might cause problems.